packages/iov-keycontrol/src/userprofile.spec.ts

```diff
@@ -37,6 +42,275 @@ describe("UserProfile", () => {
   const defaultMnemonic3 = "judge edge effort prepare caught lawn mask yellow butter phone dragon more";
   const defaultEncryptionPassword = "my super str0ng and super long password";
 
+  describe("deriveEncryptionKey", () => {
+    it("works", async () => {
+      const key = await UserProfile.deriveEncryptionKey("foobar");
+      expect(key.formatVersion).toEqual(1);
+      expect(key.data.length).toEqual(32);
+    });
+
+    it("can use a different version explicitly", async () => {
+      const key = await UserProfile.deriveEncryptionKey("foobar", 2);
+      expect(key.formatVersion).toEqual(2);
+      expect(key.data.length).toEqual(32);
+    });
+  });
+
+  describe("loadFrom", () => {
+    it("stored in and loaded from storage", async () => {
+      const db = levelup(MemDownConstructor<string, string>());
+
+      const createdAt = new ReadonlyDate("1985-04-12T23:20:50.521Z");
+      const keyring = new Keyring();
+      const original = new UserProfile({ createdAt: createdAt, keyring: keyring });
+
+      await original.storeIn(db, defaultEncryptionPassword);
+
+      const restored = await UserProfile.loadFrom(db, defaultEncryptionPassword);
+
+      expect(restored.createdAt).toEqual(original.createdAt);
+
+      await db.close();
+    });
+
+    it("stored in and loaded from storage when containing special chars", async () => {
+      const db = levelup(MemDownConstructor<string, string>());
+      const wallet1 = Ed25519HdWallet.fromMnemonic(defaultMnemonic3);
+
+      const original = new UserProfile();
+      original.addWallet(wallet1);
+      original.setWalletLabel(wallet1.id, "My secret 😛");
+
+      await original.storeIn(db, defaultEncryptionPassword);
+      const restored = await UserProfile.loadFrom(db, defaultEncryptionPassword);
+
+      expect(restored.wallets.value).toEqual(original.wallets.value);
+
+      await db.close();
+    });
+
+    it("fails when loading with wrong key", async () => {
+      const db = levelup(MemDownConstructor<string, string>());
+
+      const createdAt = new ReadonlyDate("1985-04-12T23:20:50.521Z");
+      const keyring = new Keyring();
+      const original = new UserProfile({ createdAt: createdAt, keyring: keyring });
+
+      await original.storeIn(db, defaultEncryptionPassword);
+
+      const otherEncryptionPassword = "something wrong";
+      await UserProfile.loadFrom(db, otherEncryptionPassword)
+        .then(() => fail("loading must not succeed"))
```

Comment:
What about passing two handlers to the `then` instead of `.then().catch()`? It represents the flow slightly more faithfully I think.

Reply:
I did not know that .then can take more than one argument and this is how we wrote it in many many places in this project.

But this one is also fine:

```ts
      await UserProfile.loadFrom(db, otherEncryptionPassword).then(
        () => fail("loading must not succeed"),
        error => expect(error).toMatch(/invalid usage/),
      );
```

I tend to agree that your suggestion makes sense. But maybe we better do this as a bigger refactoring through the whole project.